### DISCUSSION TITLE
Should the arguments passed to useQuery and similar hooks be referentially stable?

### DISCUSSION THREAD

**Discussion by mkarajohn** [TanStack/query#3737] [Q&A]
**DISCLAIMER**: _This was originally asked on StackOverflow (https://stackoverflow.com/questions/72723445/should-the-arguments-passed-to-react-querys-usequery-hook-be-referentially-stab) and as of the time of writing has yet to receive an answer._

I am going through an existing codebase, looking at some data fetching hooks that use [`useQuery`](https://react-query.tanstack.com/guides/queries)

I am noticing that, almost everywhere, care is taken so that the arguments passed in the `useQuery` hook (and other similar `react-query` hooks) are referentially the same between hook executions.

For example

```js
//custom_data_fetching_hook.js

import { useQuery } from 'react-query';

function fetchSomeStuff() {
  // server data request implementation
}

const queryOptions = {
 // object as described in the 3rd argument here:
 // https://react-query.tanstack.com/reference/useQuery
};

export function useSomeServerStuff() {
  return useQuery(['query key'], fetchSomeStuff, queryOptions);
}
```

My question is this:

Is it necessary to ensure referential stability for the `useQuery`, `useMutation`, etc, arguments? As far as I could tell the docs do not say you should do this. 

Would the `react-query` hooks redo calculations on every component re-render if their arguments are not referentially the same between renders? 

Would the `react-query` hooks return some referentially different items within their return object value if their arguments are not referentially the same between renders? (for example the `mutate` function inside the return object of `useMutation`)

**Comment by TkDodo** [comment]
stackoverflow question has been answered. No, arguments do not need to be referentially stable.
